plugins/UM3NetworkPrinting/tests/TestSendMaterialJob.py

```diff
@@ -0,0 +1,210 @@
+from cura.CuraApplication import CuraApplication
+from plugins.UM3NetworkPrinting.src.SendMaterialJob import SendMaterialJob
+
+
+@patch("builtins.open", lambda _, __: io.StringIO("<xml></xml>"))
+@patch("UM.MimeTypeDatabase.MimeTypeDatabase.getMimeTypeForFile",
+       lambda _: MimeType(name = "application/x-ultimaker-material-profile", comment = "Ultimaker Material Profile",
+                          suffixes = ["xml.fdm_material"]))
+@patch("UM.Resources.Resources.getAllResourcesOfType", lambda _: ["/materials/generic_pla_white.xml.fdm_material"])
+class TestSendMaterialJob(TestCase):
+    _LOCAL_MATERIAL_WHITE = {"type": "material", "status": "unknown", "id": "generic_pla_white",
+                             "base_file": "generic_pla_white", "setting_version": "5", "name": "White PLA",
+                             "brand": "Generic", "material": "PLA", "color_name": "White",
+                             "GUID": "badb0ee7-87c8-4f3f-9398-938587b67dce", "version": "1", "color_code": "#ffffff",
+                             "description": "Test PLA White", "adhesion_info": "Use glue.", "approximate_diameter": "3",
+                             "properties": {"density": "1.00", "diameter": "2.85", "weight": "750"},
+                             "definition": "fdmprinter", "compatible": True}
+
+    _LOCAL_MATERIAL_BLACK = {"type": "material", "status": "unknown", "id": "generic_pla_black",
+                             "base_file": "generic_pla_black", "setting_version": "5", "name": "Yellow CPE",
+                             "brand": "Ultimaker", "material": "CPE", "color_name": "Black",
+                             "GUID": "5fbb362a-41f9-4818-bb43-15ea6df34aa4", "version": "1", "color_code": "#000000",
+                             "description": "Test PLA Black", "adhesion_info": "Use glue.", "approximate_diameter": "3",
+                             "properties": {"density": "1.01", "diameter": "2.85", "weight": "750"},
+                             "definition": "fdmprinter", "compatible": True}
+
+    _REMOTE_MATERIAL_WHITE = {
+        "guid": "badb0ee7-87c8-4f3f-9398-938587b67dce",
+        "material": "PLA",
+        "brand": "Generic",
+        "version": 1,
+        "color": "White",
+        "density": 1.00
+    }
+
+    _REMOTE_MATERIAL_BLACK = {
+        "guid": "5fbb362a-41f9-4818-bb43-15ea6df34aa4",
+        "material": "PLA",
+        "brand": "Generic",
+        "version": 2,
+        "color": "Black",
+        "density": 1.00
+    }
+
+    @patch("plugins.UM3NetworkPrinting.src.ClusterUM3OutputDevice")
+    def test_run(self, device_mock):
+        job = SendMaterialJob(device_mock)
+        job.run()
+
+        # We expect the materials endpoint to be called when the job runs.
+        device_mock.get.assert_called_with("materials/", on_finished = job._onGetRemoteMaterials)
+
+    @patch("plugins.UM3NetworkPrinting.src.ClusterUM3OutputDevice")
+    @patch("PyQt5.QtNetwork.QNetworkReply")
```

Comment:
the mocks needed in every method might as well be in the class level

Reply:
They are